trio/_highlevel_generic.py

```diff
@@ -71,29 +71,42 @@ class StapledStream(HalfCloseableStream):
     """
     send_stream = attr.ib()
     receive_stream = attr.ib()
+    # This is needed to correctly handle the case where send_eof calls aclose,
+    # because aclose interrupts other tasks, while send_eof is supposed to
+    # error out if there are other tasks.
+    _send_conflict_detector = attr.ib(
+        init=False,
+        factory=lambda:
+        ConflictDetector("another task is sending on this stream"),
+    )
 
     async def send_all(self, data):
         """Calls ``self.send_stream.send_all``.
 
         """
-        return await self.send_stream.send_all(data)
+        with self._send_conflict_detector:
+            return await self.send_stream.send_all(data)
 
     async def wait_send_all_might_not_block(self):
         """Calls ``self.send_stream.wait_send_all_might_not_block``.
 
         """
-        return await self.send_stream.wait_send_all_might_not_block()
+        with self._send_conflict_detector:
+            return await self.send_stream.wait_send_all_might_not_block()
 
     async def send_eof(self):
         """Shuts down the send side of the stream.
 
-        If ``self.send_stream.send_eof`` exists, then calls it. Otherwise,
-        calls ``self.send_stream.aclose()``.
+        If ``self.send_stream.send_eof`` exists, then attempts to call it.
+        Otherwise, calls ``self.send_stream.aclose()``.
 
         """
-        if hasattr(self.send_stream, "send_eof"):
-            return await self.send_stream.send_eof()
-        else:
+        with self._send_conflict_detector:
+            if hasattr(self.send_stream, "send_eof"):
```

Comment:
now that all classes inheriting from `Stream` are required to implement `send_eof`, why do we need to check that the sending stream has the `send_eof` method here? i think i am probably misunderstanding a subtle point here. maybe this is for backwards compatibility or maybe there are other types of streams that i am unaware of, which can be "stapled" together.

Reply:
All `Stream`s have to implement `send_eof`, but unidirectional `SendStream`s don't have to, and in this method, all we know about `self.send_stream` is that it's a `SendStream`, not necessarily a `Stream`.

I'm not actually sure if this opportunistic attempt to call `send_eof` accomplishes anything useful... it would be simpler, and maybe better, to unconditionally call `aclose`. But I decided this PR was already doing enough, so I'd leave worrying about that to another time...